### PR TITLE
[DNM] RFC: LwM2M Bootstrap Feature Support

### PR DESCRIPTION
**Pull request by mike-scott**
This series of patches adds LwM2M Bootstrap support to the Zephyr LwM2M client.

### CODE REVIEW

**Comment by mike-scott** [comment]
This is a rather large patch set that I need to probably clean up a bit more prior to merging.  However, it also needs to be tested across several servers, so I'm posting it DNM / RFC for now.

@therealprof feel free to review / comment

NOTE:
Once this is merged, my focus will shift to:
1. Addressing current issues w/ LwM2M client such as backoff timers during error paths, etc.
2. Re-writing the current LwM2M library to use the sockets API

**Comment by mike-scott** [comment]
Specifically, https://github.com/zephyrproject-rtos/zephyr/pull/8570/commits/07cb4ece1d254126e603c722fb4b39acbc3759fa commit needs to be broken down into smaller commits.

**Comment by mike-scott** [comment]
@dbkinder No need to review for doc changes yet.

**Comment by mike-scott** [comment]
@jukkar currently, I'm abusing a NET_APP config:
CONFIG_NET_APP_PEER_IPV6_ADDR with a value of "coap[s]://[2001:db8::2]:568[3|4]"

This is not the intended use for this config.  I'd like a recommendation on options that handle the following data:
1. IP address
2. Protocol [coap|coaps]
3. Conection Port

This could all be determined in the sample and then passed on as a single string to the LwM2M client.

**Comment by therealprof** [comment]
Looking forward to testing that tomorrow or so.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 1]
```diff
@@ -117,16 +114,20 @@ static void set_sm_state(u8_t sm_state)
...
 	/* Determine if a callback to the app is needed */
-	if (sm_state == ENGINE_BOOTSTRAP_DONE) {
-		event = LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_COMPLETE;
-	} else if (client.engine_state == ENGINE_UPDATE_SENT &&
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	if (sm_state == ENGINE_BOOTSTRAP_REG_DONE) {
+		event = LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_REG_COMPLETE;
+	} else if (sm_state == ENGINE_BOOTSTRAP_TRANS_DONE) {
+		event = LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_TRANSFER_COMPLETE;
+#endif
```
Comment (therealprof) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c]: Missing closing }

**Review by therealprof — CHANGES_REQUESTED** [review]
Even with that compile fix it doesn't work for me since it is going into an endless loop:

```
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/0/1, buf:0x20007597, buflen:1
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:1/0/1, buf:0x20004160, buflen:4
[lib/lwm2m_rd_client] [DBG] sm_select_next_sec_inst: sec_obj_inst: 0
[lib/lwm2m_rd_client] [INF] sm_do_bootstrap_reg: Bootstrap started with endpoint 'frdm_k64f' with client lifetime 30
[lib/lwm2m_engine] [DBG] lwm2m_engine_start: URL: coap://212.18.24.70
[lib/lwm2m_rd_client] [DBG] sm_do_bootstrap_reg: Register ID with bootstrap server [212.18.24.70] as 'ep=frdm_k64f'
[lib/lwm2m_engine] [DBG] lwm2m_udp_receive: checking for reply from [212.18.24.70]
[lib/lwm2m_rd_client] [DBG] do_bootstrap_reply_cb: Bootstrap callback (code:4.4)
[lib/lwm2m_rd_client] [ERR] do_bootstrap_reply_cb: Failed: NOT_FOUND.  Not Retrying.
[lib/lwm2m_engine] [DBG] lwm2m_udp_receive: reply 0x200000bc handled and removed
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/0/1, buf:0x20007597, buflen:1
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/0/1, buf:0x20007597, buflen:1
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/0/1, buf:0x20007597, buflen:1
...
```

Will need to have a closer look.

**Comment by therealprof** [comment]
The bootstrap also happily accepts being sent to a coaps registration server without DTLS enabled in the build, again causing an endless loop:
```
[lwm2m_obj_server] [DBG] server_create: Create LWM2M server instance: 1
[lib/lwm2m_engine] [DBG] lwm2m_udp_receive: checking for reply from [212.18.24.70]
[lib/lwm2m_rd_client] [INF] engine_bootstrap_finish: Bootstrap data transfer done!
[lwm2m-client] [DBG] rd_client_event: Bootstrap transfer complete
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/1/1, buf:0x20007597, buflen:1
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:1/1/1, buf:0x20004160, buflen:4
[lib/lwm2m_rd_client] [DBG] sm_select_next_sec_inst: sec_obj_inst: 1
[lib/lwm2m_rd_client] [INF] sm_do_registration: RD Client started with endpoint 'frdm_k64f' with client lifetime 60
[lib/lwm2m_rd_client] [ERR] sm_do_registration: Cannot init LWM2M engine (-43)
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:0/1/1, buf:0x20007597, buflen:1
[lib/lwm2m_engine] [DBG] lwm2m_engine_get: path:1/1/1, buf:0x20004160, buflen:4
[lib/lwm2m_rd_client] [DBG] sm_select_next_sec_inst: sec_obj_inst: 1
...
```

**Comment by therealprof** [comment]
... enabling DTLS in the build and bootstrapping without DTLS to a server with DTLS causes an exception:
```
[lwm2m-client] [DBG] rd_client_event: Bootstrap transfer complete
[lib/lwm2m_engine] [DBG] lwm2m_engine_context_close: Releasing DTLS context 0x20004814
***** USAGE FAULT *****
  Executing thread ID (thread): 0x20004e4c
  Faulting instruction address:  0x0
  Illegal use of the EPSR
Fatal fault in thread 0x20004e4c! Aborting.
```

Need to bring out the heavy gear I guess. 😉

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 1 above)
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c]: Great find.  I wonder if I pulled in the wrong patches some how.. this never would have compiled when I was testing.

**Comment by mike-scott** [comment]
@therealprof Don't spend time debugging this yet, let me verify that this is the same patch set I've been testing locally.
NOTE: Your comments apply tho, as there is very little checking to make sure the user is doing something that "makes sense".

**Comment by therealprof** [comment]
@mike-scott Any update? Maybe we can split up the patch and have some items merged as is like the create/delete callback support.

**Comment by mike-scott** [comment]
@therealprof busy week and now holiday.  I'll split the top 4 off into a separate PR.  Then try and fix #5 in a better way in a new PR.

**Comment by nashif** [comment]
@mike-scott status of this PR?

**Comment by mike-scott** [comment]
@nashif It's going to be reworked a bit once the socket support goes in.

I'm not sure if @therealprof ever found it useful.  I put it up mostly for testing.

Maybe keep it open a bit longer.  I've been context switching too much lately.

**Comment by therealprof** [comment]
@mike-scott Very useful. I'm still hopeful for the split version of the patches. At the moment it's a whee bit buggy but that's probably because there's too much stuff at once.

**Comment by mike-scott** [comment]
Closing stale PR.  These changes are included in: https://github.com/zephyrproject-rtos/zephyr/pull/9832

**Pull request closed without merging**